# EasyOCR 인식 결과 테스트

데이터셋 : 공공행정문서 OCR (2001 폴더 - 200개의 파일)

실행 환경 : 실행 시간 절약을 위해 CUDA, cuDNN를 설치하여 GPU 환경에서 실행

## 라벨링 데이터(JSON) 파싱

In [9]:
import json

with open("./5350109-2001-0001-0001.json", encoding="UTF8") as file:
    data = json.load(file)
    
    json_test = data['annotations']
    for k in json_test:
        print(k['annotation.text'])

"김해를
아름답게
시민을
행복하게"
김
해
시
우
621-010/
경남
김해시
부원동
623
번지/
전화(055)330-3363
/전송330-3689
허가민원과
과장
류보현
공장설립담당주사
감경석
담당자
이창수
http://gimhae.gc.kr
E-mail
:
lcs@mail.kimhae.kyongnam.kr
문서번호
허가55142 -
4043
보존기간
시
장
시행일자
2001.
5.
4
(
년)
공개여부
4
(경
유)
부시장
수
신
부산시
사상구
감전동
159-5번지
실
장
동일중공업(주)대표
담당주사
과
장
전결
참
조
기
안
심사자
임삼성
목
공장설립
변경승인
통보
1.
귀하께서
제출한
공장설립변경승인신청에
대하여
공업배치
및
공장설립에
관한법률
제13조
제3항의
규정에
의거
붙임과
같이
승인서를
교부하오니,
조건사항
이행에
만전을
기하시기
바랍니다.
○
공장설립
변경승인
내용
회사명
소재지
대표자
업
종
(분류번호)
한림.안하
선박구성부분품
당초
영광산업사
1110-3
제조업
(35114)
승인
외1필지
변경
동일중공업
한림.안하
1110-3
승인
(주)
외1필지
제조업
(29169)
변경사유
:
회사명(대표자)
업종,
건축면적
변경
2.
공장설립승인
및
농지전용허가에
따른
지방세법
제161조
및
동법시행령
제124조의
규정에
의한
면허세(공장
:
8,000원,
농지
18,000원)를
납부하여야
합니다.
3.
건축물
사용승인을
얻고
기계.장치의
설치를
완료한
후
2개월이내에
공장
설립완료신고를
하여야
합니다.
이를
위반하는
경우에는
200만원이하의
과태료처분을
받게되오니
이점유념하시기
바랍니다.
4.
우리시는
공장을
설립하는
업체에
대하여
특별한
사유가
없는
한
본사를
우리시에
소재하도록
유도하고
있습니다.
귀사의
본점이
우리시내에
계속
유지되기를
당부드립니다.
붙임
:
공장설립
변경승인서
1부
끝.
김
해
제
구분
(단위
:
㎡
)
용지면적
건축면적
부대
(제조시설)
시설
3,204.0
495.0
3,204.0
494.2
754.12
시
장
46
기

## 인식 결과만 추출 (슬라이싱)

In [19]:
import numpy as np
import cv2
import random
import matplotlib.pyplot as plt
from PIL import ImageFont, ImageDraw, Image
import easyocr

reader = easyocr.Reader(['ko','en'], gpu=True)
result = reader.readtext("./5350109-2001-0001-0001.jpg", width_ths=0.1, adjust_contrast=0.1)

for i in result:
    print(i[1])

CUDA not available - defaulting to CPU. Note: This module is much faster with a GPU.


"김해틀
아름답게
시민올
행복하게"
김
해
우
621-010/
경남
김해시
부원동
623
번지/
전화(055)330-3363
/전송330-3689
허가민원과
과 장 류 보 현
공장설립담당주사
감 경 석
담당자
이 창 수
http:llgimhae.go.kr
E-mail
Ics@mail kimhae kyongnamkr
문서번호
허가55142
4043
보존기간
장
시행일자
2O01.
4 ( 년)
공개여부
(경   유)
부 시 장
발외함
수   신
부산시
사상구
감전동
159-5번지
실
장
동일중공업주)
대표
담당주사
과   장 |
전   결
6y
참   조
기   안
#@이3쥐
양남
심사자
제   목
공장설립
변경승인
통보
1.
귀하께서
제출한
공장설립변경승인신청에
대하여
공업배치
및
공장설립에
관한법률
제13조
제3항의
규정에
의거
불임과
같이
승인서률
교부하오니,
조건사항
이행에
만전올
기하시기
바랍니다.
공장설립
변경승인
내용
단위
mi'
)
업   종
건축면적
부대
구분
회사명
소재지
대표자
용지면적
(분류번호)
(제조시설)
시설
당초
한림.안하
선박구성부분품
영광산업사
1110-3
3,204.0
495.0
승인
외1 필지
제조업
(35114)
변경
동일중공업
한림.안하
기타물품취 {장비
1110-3
3,204.0
494.2
754.12
승인
(주)
외1 필지
제조업
(29169)
변경사유
회사명(대표자)
업종,
건축면적
변경
2.
공장설립승인
및
농지전용허가에
따른
지방세법
제161조
및
동법시 행령
제124조의
규정에
의한
면혀 세(공장
8,000원,
농지
18,000원)틀
납부하여야
합니다.
3.
건축물
사용승인올
얻고
기계 장치의
설치률
완료한
후
2개월이내에
공장
설립완료신고룹
하여야
합니다.
이틀
위반하는
경우에는
200만원이하의
과태료처분올
밭게되오니
이점유넘하시기
바람니다.
4.
우리시논
공장올
설립하는
업체에
대하여
특별한
사유가
없는
한
본사을
우리시에
소재하도록
유도하고
있습니다.
귀사의
본점이
우리시내에
계속
유지되기
당부드럽니다.
불임

## 인식 결과 비교

In [28]:
import numpy as np
import cv2
import random
import matplotlib.pyplot as plt
from PIL import ImageFont, ImageDraw, Image
import easyocr
import json

# easyocr 인식 실행
reader = easyocr.Reader(['ko','en'], gpu=True)
result = reader.readtext("./5350109-2001-0001-0001.jpg", width_ths=0.1, adjust_contrast=0.1)

# 인식 결과 출력
print("<OCR 인식 결과>")
for i in result:
    print(i[1])
    
print("")
print("####################################################################################################")
print("")

# 라벨링 데이터(json) 파일 불러오기
with open("./5350109-2001-0001-0001.json", encoding="UTF8") as file:
    data = json.load(file)
    json_test = data['annotations']
    
    # 문자 값만 출력
    print("<라벨링 데이터>")
    for k in json_test:
        print(k['annotation.text'])
    
    # 인식 결과값과 라벨링 데이터 값 추출
    result_texts = {i[1] for i in result}
    json_test_texts = {k['annotation.text'] for k in json_test}
    
    # 일치 비율 계산
    matching_texts = result_texts.intersection(json_test_texts)
    percentage = (len(matching_texts) / len(result_texts)) * 100
    
    print("")
    print("####################################################################################################")
    print("")
    
    # 일치율 및 정확도 출력
    if matching_texts:
        print("인식 성공 : ", matching_texts)
        print("")
        print("####################################################################################################")
        print("")
        print("정확도 : {:.2f}%".format(percentage))
    else:
        print("정확한 인식 값이 없습니다.")

CUDA not available - defaulting to CPU. Note: This module is much faster with a GPU.


<OCR 인식 결과>
"김해틀
아름답게
시민올
행복하게"
김
해
우
621-010/
경남
김해시
부원동
623
번지/
전화(055)330-3363
/전송330-3689
허가민원과
과 장 류 보 현
공장설립담당주사
감 경 석
담당자
이 창 수
http:llgimhae.go.kr
E-mail
Ics@mail kimhae kyongnamkr
문서번호
허가55142
4043
보존기간
장
시행일자
2O01.
4 ( 년)
공개여부
(경   유)
부 시 장
발외함
수   신
부산시
사상구
감전동
159-5번지
실
장
동일중공업주)
대표
담당주사
과   장 |
전   결
6y
참   조
기   안
#@이3쥐
양남
심사자
제   목
공장설립
변경승인
통보
1.
귀하께서
제출한
공장설립변경승인신청에
대하여
공업배치
및
공장설립에
관한법률
제13조
제3항의
규정에
의거
불임과
같이
승인서률
교부하오니,
조건사항
이행에
만전올
기하시기
바랍니다.
공장설립
변경승인
내용
단위
mi'
)
업   종
건축면적
부대
구분
회사명
소재지
대표자
용지면적
(분류번호)
(제조시설)
시설
당초
한림.안하
선박구성부분품
영광산업사
1110-3
3,204.0
495.0
승인
외1 필지
제조업
(35114)
변경
동일중공업
한림.안하
기타물품취 {장비
1110-3
3,204.0
494.2
754.12
승인
(주)
외1 필지
제조업
(29169)
변경사유
회사명(대표자)
업종,
건축면적
변경
2.
공장설립승인
및
농지전용허가에
따른
지방세법
제161조
및
동법시 행령
제124조의
규정에
의한
면혀 세(공장
8,000원,
농지
18,000원)틀
납부하여야
합니다.
3.
건축물
사용승인올
얻고
기계 장치의
설치률
완료한
후
2개월이내에
공장
설립완료신고룹
하여야
합니다.
이틀
위반하는
경우에는
200만원이하의
과태료처분올
밭게되오니
이점유넘하시기
바람니다.
4.
우리시논
공장올
설립하는
업체에
대하여
특별한
사유가
없는
한
본사을
우리시에
소재하도록
유도하고
있습니다.
귀사의
본점이
우리시내에
계속
유지되

순서 상관없이 단순 배열 값 비교여서 정확도 계산에서 오차가 존재하는 것으로 예상

## 인식 결과 비교 (5개의 문서만 실행)

In [48]:
import numpy as np
import cv2
import random
import matplotlib.pyplot as plt
from PIL import ImageFont, ImageDraw, Image
import easyocr
import json
import os
import time

# 폴더 경로
folder_path_source = '2001_source'
folder_path_label = '2001_label'

# 폴더 안에 있는 모든 파일을 리스트 형태로 가져오기
file_list = os.listdir(folder_path_source)

# 파일 경로와 파일 이름 결합
file_paths = [os.path.join(folder_path_source, file_name) for file_name in file_list]

# 인식 시간 고려하여 파일 5개만 선택하여 실행
file_paths = file_paths[:5]

# easyocr 모델 로드
reader = easyocr.Reader(['ko','en'], gpu=True)

# 파일 수와 평균 경과 시간, 평균 정확도를 저장할 변수 초기화
file_count = len(file_paths)
total_elapsed_time = 0
total_accuracy = 0

# 각 파일에 대해 OCR 수행
for file_path in file_paths:
    filename = os.path.splitext(os.path.basename(file_path))[0]
    print("파일명 :", os.path.basename(file_path))
    
    # 시작 시간 저장
    start_time = time.time()
    
    # OCR 인식 결과 추출
    result = reader.readtext(folder_path_source+"/"+filename+".jpg", width_ths=0.1, adjust_contrast=0.1)
    
    # 인식 경과 시간 계산
    elapsed_time = time.time() - start_time
    
    # 총 경과 시간 계산
    total_elapsed_time += elapsed_time
    
    # 문자 인식 결과값 출력
    print("<OCR 인식 결과>")
    for i in result:
        print(i[1])

    print("####################################################################################################")
    
    # 라벨링 데이터(json) 파일 불러오기
    with open(folder_path_label+"/"+filename+".json", encoding="UTF8") as file:
        data = json.load(file)
        json_test = data['annotations']
    
        # 문자 값만 출력
        print("<라벨링 데이터>")
        for k in json_test:
            print(k['annotation.text'])
    
        # 인식 결과값과 라벨링 데이터 값 추출
        result_texts = {i[1] for i in result}
        json_test_texts = {k['annotation.text'] for k in json_test}
    
        # 일치 비율 계산
        matching_texts = result_texts.intersection(json_test_texts)
        percentage = (len(matching_texts) / len(result_texts)) * 100
    
        print("####################################################################################################")
    
        # 일치율 및 정확도, 경과 시간 출력
        if matching_texts:
            print("인식 성공 : ", matching_texts)
            print("####################################################################################################")
            print("정확도 : {:.2f}%".format(percentage))
            print("경과 시간 : {:.3f}초".format(elapsed_time))
            total_accuracy += percentage
        else:
            print("정확한 인식 값이 없습니다.")
            
        print("")
        print("####################################################################################################")
        print("####################################################################################################")
        print("####################################################################################################")
        print("")
        
# 평균 경과 시간과 평균 정확도 계산 및 출력
average_elapsed_time = total_elapsed_time / file_count
average_accuracy = total_accuracy / file_count
print("평균 경과 시간 : {:.3f}초".format(average_elapsed_time))
print("평균 정확도 : {:.2f}%".format(average_accuracy))

CUDA not available - defaulting to CPU. Note: This module is much faster with a GPU.


파일명 : 5350109-2001-0001-0001.jpg
<OCR 인식 결과>
"김해틀
아름답게
시민올
행복하게"
김
해
우
621-010/
경남
김해시
부원동
623
번지/
전화(055)330-3363
/전송330-3689
허가민원과
과 장 류 보 현
공장설립담당주사
감 경 석
담당자
이 창 수
http:llgimhae.go.kr
E-mail
Ics@mail kimhae kyongnamkr
문서번호
허가55142
4043
보존기간
장
시행일자
2O01.
4 ( 년)
공개여부
(경   유)
부 시 장
발외함
수   신
부산시
사상구
감전동
159-5번지
실
장
동일중공업주)
대표
담당주사
과   장 |
전   결
6y
참   조
기   안
#@이3쥐
양남
심사자
제   목
공장설립
변경승인
통보
1.
귀하께서
제출한
공장설립변경승인신청에
대하여
공업배치
및
공장설립에
관한법률
제13조
제3항의
규정에
의거
불임과
같이
승인서률
교부하오니,
조건사항
이행에
만전올
기하시기
바랍니다.
공장설립
변경승인
내용
단위
mi'
)
업   종
건축면적
부대
구분
회사명
소재지
대표자
용지면적
(분류번호)
(제조시설)
시설
당초
한림.안하
선박구성부분품
영광산업사
1110-3
3,204.0
495.0
승인
외1 필지
제조업
(35114)
변경
동일중공업
한림.안하
기타물품취 {장비
1110-3
3,204.0
494.2
754.12
승인
(주)
외1 필지
제조업
(29169)
변경사유
회사명(대표자)
업종,
건축면적
변경
2.
공장설립승인
및
농지전용허가에
따른
지방세법
제161조
및
동법시 행령
제124조의
규정에
의한
면혀 세(공장
8,000원,
농지
18,000원)틀
납부하여야
합니다.
3.
건축물
사용승인올
얻고
기계 장치의
설치률
완료한
후
2개월이내에
공장
설립완료신고룹
하여야
합니다.
이틀
위반하는
경우에는
200만원이하의
과태료처분올
밭게되오니
이점유넘하시기
바람니다.
4.
우리시논
공장올
설립하는
업체에
대하여
특별한
사유가
없는
한
본사을
우리시에
소재하도

## 인식 결과 비교 (5개만 / 결과값만 출력)

In [2]:
import numpy as np
import cv2
import random
import matplotlib.pyplot as plt
from PIL import ImageFont, ImageDraw, Image
import easyocr
import json
import os
import time

# 폴더 경로
folder_path_source = '2001_source'
folder_path_label = '2001_label'

# 폴더 안에 있는 모든 파일을 리스트 형태로 가져오기
file_list = os.listdir(folder_path_source)

# 파일 경로와 파일 이름 결합
file_paths = [os.path.join(folder_path_source, file_name) for file_name in file_list]

# 인식 시간 고려하여 파일 5개만 선택하여 실행
file_paths = file_paths[:5]

# easyocr 모델 로드
reader = easyocr.Reader(['ko','en'], gpu=True)

# 파일 수와 평균 경과 시간, 평균 정확도를 저장할 변수 초기화
file_count = len(file_paths)
total_elapsed_time = 0
total_accuracy = 0

# 각 파일에 대해 OCR 수행
for file_path in file_paths:
    filename = os.path.splitext(os.path.basename(file_path))[0]
    print("파일명 :", os.path.basename(file_path))
    print("인식 실행 중")
    
    # 시작 시간 저장
    start_time = time.time()
    
    # OCR 인식 결과 추출
    result = reader.readtext(folder_path_source+"/"+filename+".jpg", width_ths=0.1, adjust_contrast=0.1)
    
    # 인식 경과 시간 계산
    elapsed_time = time.time() - start_time
    
    # 총 경과 시간 계산
    total_elapsed_time += elapsed_time
    
    '''
    # 문자 인식 결과값 출력
    print("<OCR 인식 결과>")
    for i in result:
        print(i[1])

    print("####################################################################################################")
    '''
    
    # 라벨링 데이터(json) 파일 불러오기
    with open(folder_path_label+"/"+filename+".json", encoding="UTF8") as file:
        data = json.load(file)
        json_test = data['annotations']
        
        '''
        # 문자 값만 출력
        print("<라벨링 데이터>")
        for k in json_test:
            print(k['annotation.text'])
            
        print("####################################################################################################")
        '''
    
        # 인식 결과값과 라벨링 데이터 값 추출
        result_texts = {i[1] for i in result}
        json_test_texts = {k['annotation.text'] for k in json_test}
    
        # 일치 비율 계산
        matching_texts = result_texts.intersection(json_test_texts)
        percentage = (len(matching_texts) / len(result_texts)) * 100
    
        # 일치율 및 정확도, 경과 시간 출력
        if matching_texts:
            
            '''
            print("인식 성공 : ", matching_texts)
            print("####################################################################################################")
            '''
            
            print("정확도 : {:.2f}%".format(percentage))
            print("경과 시간 : {:.3f}초".format(elapsed_time))
            total_accuracy += percentage
        else:
            print("정확한 인식 값이 없습니다.")
            
        print("")
        
        '''
        print("####################################################################################################")
        print("####################################################################################################")
        print("####################################################################################################")
        print("")
        '''

# 평균 경과 시간과 평균 정확도 계산 및 출력
average_elapsed_time = total_elapsed_time / file_count
average_accuracy = total_accuracy / file_count
print("평균 경과 시간 : {:.3f}초".format(average_elapsed_time))
print("평균 정확도 : {:.2f}%".format(average_accuracy))

파일명 : 5350109-2001-0001-0001.jpg
인식 실행 중
정확도 : 69.71%
경과 시간 : 2.714초

파일명 : 5350109-2001-0001-0002.jpg
인식 실행 중
정확도 : 58.82%
경과 시간 : 1.070초

파일명 : 5350109-2001-0001-0004.jpg
인식 실행 중
정확도 : 58.87%
경과 시간 : 2.011초

파일명 : 5350109-2001-0001-0005.jpg
인식 실행 중
정확도 : 59.12%
경과 시간 : 2.149초

파일명 : 5350109-2001-0001-0006.jpg
인식 실행 중
정확도 : 54.76%
경과 시간 : 1.028초

평균 경과 시간 : 1.795초
평균 정확도 : 60.26%


# [최종 결과]

## 인식 결과 비교 (모든 데이터 / 결과값만 출력)

### EasyOCR 속성 값 기본으로 실행

width_ths=0.1 / adjust_contrast=0.1

In [3]:
import easyocr
import json
import os
import time

# 폴더 경로
folder_path_source = '2001_source'
folder_path_label = '2001_label'

# 폴더 안에 있는 모든 파일을 리스트 형태로 가져오기
file_list = os.listdir(folder_path_source)

# 파일 경로와 파일 이름 결합
file_paths = [os.path.join(folder_path_source, file_name) for file_name in file_list]

# 파일 수와 평균 경과 시간, 평균 정확도를 저장할 변수 초기화
file_count = len(file_paths)
total_elapsed_time = 0
total_accuracy = 0

'''
# 파일 개수 지정 (ex) 5개)
file_paths = file_paths[:5]
file_count = 5
'''

# easyocr 모델 로드
reader = easyocr.Reader(['ko','en'], gpu=True)

# 각 파일에 대해 OCR 수행
for file_path in file_paths:
    filename_split = os.path.splitext(os.path.basename(file_path))[0]
    print("파일명 :", os.path.basename(file_path))
    print("인식 실행 중")
    
    # 시작 시간 저장
    start_time = time.time()
    
    # OCR 인식 결과 추출
    result = reader.readtext(folder_path_source+"/"+filename_split+".jpg", width_ths=0.1, adjust_contrast=0.1)
    
    # 인식 경과 시간 계산
    elapsed_time = time.time() - start_time
    
    # 총 경과 시간 계산
    total_elapsed_time += elapsed_time
    
    '''
    # 문자 인식 결과값 출력
    print("<OCR 인식 결과>")
    for i in result:
        print(i[1])
    '''
    
    # 라벨링 데이터(json) 파일 불러오기
    with open(folder_path_label+"/"+filename_split+".json", encoding="UTF8") as file:
        data = json.load(file)
        json_test = data['annotations']
        
        '''
        # 문자 값만 출력
        print("<라벨링 데이터>")
        for k in json_test:
            print(k['annotation.text'])
        '''
    
        # 인식 결과값과 라벨링 데이터 값 추출
        result_texts = {i[1] for i in result}
        json_test_texts = {k['annotation.text'] for k in json_test}
    
        # 일치 비율 계산
        matching_texts = result_texts.intersection(json_test_texts)
        percentage = (len(matching_texts) / len(result_texts)) * 100
    
        # 일치율 및 정확도, 경과 시간 출력
        if matching_texts:
            
            '''
            print("인식 성공 : ", matching_texts)
            '''
            
            print("정확도 : {:.2f}%".format(percentage))
            print("경과 시간 : {:.2f}초".format(elapsed_time))
            total_accuracy += percentage
        else:
            print("정확한 인식 값이 없습니다.")
            
        print("")

# 총 데이터 수 및 총 소요시간, 평균 경과 시간, 평균 정확도 계산
average_elapsed_time = total_elapsed_time / file_count
average_accuracy = total_accuracy / file_count
total_minutes = int(total_elapsed_time // 60)
total_seconds = int(total_elapsed_time % 60)

# 테스트 결과 출력
print("총 데이터 수 : "+str(file_count)+"개")
print("총 소요 시간 : "+str(total_minutes)+"분 "+str(total_seconds)+"초")
print("평균 인식 시간 : {:.2f}초".format(average_elapsed_time))
print("평균 정확도 : {:.2f}%".format(average_accuracy))

파일명 : 5350109-2001-0001-0001.jpg
인식 실행 중
정확도 : 69.71%
경과 시간 : 2.69초

파일명 : 5350109-2001-0001-0002.jpg
인식 실행 중
정확도 : 58.82%
경과 시간 : 1.08초

파일명 : 5350109-2001-0001-0004.jpg
인식 실행 중
정확도 : 58.87%
경과 시간 : 2.08초

파일명 : 5350109-2001-0001-0005.jpg
인식 실행 중
정확도 : 59.12%
경과 시간 : 2.22초

파일명 : 5350109-2001-0001-0006.jpg
인식 실행 중
정확도 : 54.76%
경과 시간 : 1.00초

파일명 : 5350109-2001-0001-0009.jpg
인식 실행 중
정확도 : 52.94%
경과 시간 : 2.02초

파일명 : 5350109-2001-0001-0010.jpg
인식 실행 중
정확도 : 67.29%
경과 시간 : 1.88초

파일명 : 5350109-2001-0001-0015.jpg
인식 실행 중
정확도 : 60.47%
경과 시간 : 2.12초

파일명 : 5350109-2001-0001-0016.jpg
인식 실행 중
정확도 : 47.06%
경과 시간 : 0.67초

파일명 : 5350109-2001-0001-0017.jpg
인식 실행 중
정확도 : 57.82%
경과 시간 : 2.33초

파일명 : 5350109-2001-0001-0018.jpg
인식 실행 중
정확도 : 60.74%
경과 시간 : 2.18초

파일명 : 5350109-2001-0001-0019.jpg
인식 실행 중
정확도 : 64.71%
경과 시간 : 0.74초

파일명 : 5350109-2001-0001-0020.jpg
인식 실행 중
정확도 : 61.88%
경과 시간 : 2.94초

파일명 : 5350109-2001-0001-0021.jpg
인식 실행 중
정확도 : 49.44%
경과 시간 : 1.65초

파일명 : 5350109-2001-0001-0022.jpg
인

정확도 : 57.89%
경과 시간 : 0.98초

파일명 : 5350109-2001-0001-0466.jpg
인식 실행 중
정확도 : 67.63%
경과 시간 : 2.35초

파일명 : 5350109-2001-0001-0467.jpg
인식 실행 중
정확도 : 50.31%
경과 시간 : 2.55초

파일명 : 5350109-2001-0001-0468.jpg
인식 실행 중
정확도 : 58.14%
경과 시간 : 0.98초

파일명 : 5350109-2001-0001-0469.jpg
인식 실행 중
정확도 : 63.85%
경과 시간 : 2.02초

파일명 : 5350109-2001-0001-0470.jpg
인식 실행 중
정확도 : 61.70%
경과 시간 : 2.11초

파일명 : 5350109-2001-0001-0471.jpg
인식 실행 중
정확도 : 57.14%
경과 시간 : 0.63초

파일명 : 5350109-2001-0001-0472.jpg
인식 실행 중
정확도 : 55.19%
경과 시간 : 2.30초

파일명 : 5350109-2001-0001-0473.jpg
인식 실행 중
정확도 : 58.70%
경과 시간 : 1.01초

파일명 : 5350109-2001-0001-0479.jpg
인식 실행 중
정확도 : 58.90%
경과 시간 : 2.28초

파일명 : 5350109-2001-0001-0480.jpg
인식 실행 중
정확도 : 56.10%
경과 시간 : 0.95초

파일명 : 5350109-2001-0001-0481.jpg
인식 실행 중
정확도 : 61.31%
경과 시간 : 2.33초

파일명 : 5350109-2001-0001-0482.jpg
인식 실행 중
정확도 : 62.50%
경과 시간 : 2.20초

파일명 : 5350109-2001-0001-0483.jpg
인식 실행 중
정확도 : 56.25%
경과 시간 : 0.67초

파일명 : 5350109-2001-0001-0484.jpg
인식 실행 중
정확도 : 66.67%
경과 시간 : 1.99초

파일명 : 

### EasyOCR 속성 변경 (다운그레이드)

[변경 전] width_ths=0.1 / adjust_contrast=0.1

[변경 후] width_ths=0.1 / adjust_contrast=1 / low_text=0.7

In [4]:
import easyocr
import json
import os
import time

# 폴더 경로
folder_path_source = '2001_source'
folder_path_label = '2001_label'

# 폴더 안에 있는 모든 파일을 리스트 형태로 가져오기
file_list = os.listdir(folder_path_source)

# 파일 경로와 파일 이름 결합
file_paths = [os.path.join(folder_path_source, file_name) for file_name in file_list]

# 파일 수와 평균 경과 시간, 평균 정확도를 저장할 변수 초기화
file_count = len(file_paths)
total_elapsed_time = 0
total_accuracy = 0

'''
# 파일 개수 지정 (ex) 5개)
file_paths = file_paths[:5]
file_count = 5
'''

# easyocr 모델 로드
reader = easyocr.Reader(['ko','en'], gpu=True)

# 각 파일에 대해 OCR 수행
for file_path in file_paths:
    filename_split = os.path.splitext(os.path.basename(file_path))[0]
    print("파일명 :", os.path.basename(file_path))
    print("인식 실행 중")
    
    # 시작 시간 저장
    start_time = time.time()
    
    # OCR 인식 결과 추출 (속성 값 변경-다운그레이드)
    result = reader.readtext(folder_path_source+"/"+filename_split+".jpg", width_ths=0.1, adjust_contrast=1, low_text=0.7)
    
    # 인식 경과 시간 계산
    elapsed_time = time.time() - start_time
    
    # 총 경과 시간 계산
    total_elapsed_time += elapsed_time
    
    '''
    # 문자 인식 결과값 출력
    print("<OCR 인식 결과>")
    for i in result:
        print(i[1])
    '''
    
    # 라벨링 데이터(json) 파일 불러오기
    with open(folder_path_label+"/"+filename_split+".json", encoding="UTF8") as file:
        data = json.load(file)
        json_test = data['annotations']
        
        '''
        # 문자 값만 출력
        print("<라벨링 데이터>")
        for k in json_test:
            print(k['annotation.text'])
        '''
    
        # 인식 결과값과 라벨링 데이터 값 추출
        result_texts = {i[1] for i in result}
        json_test_texts = {k['annotation.text'] for k in json_test}
    
        # 일치 비율 계산
        matching_texts = result_texts.intersection(json_test_texts)
        percentage = (len(matching_texts) / len(result_texts)) * 100
    
        # 일치율 및 정확도, 경과 시간 출력
        if matching_texts:
            
            '''
            print("인식 성공 : ", matching_texts)
            '''
            
            print("정확도 : {:.2f}%".format(percentage))
            print("경과 시간 : {:.2f}초".format(elapsed_time))
            total_accuracy += percentage
        else:
            print("정확한 인식 값이 없습니다.")
            
        print("")

# 총 데이터 수 및 총 소요시간, 평균 경과 시간, 평균 정확도 계산
average_elapsed_time = total_elapsed_time / file_count
average_accuracy = total_accuracy / file_count
total_minutes = int(total_elapsed_time // 60)
total_seconds = int(total_elapsed_time % 60)

# 테스트 결과 출력
print("총 데이터 수 : "+str(file_count)+"개")
print("총 소요 시간 : "+str(total_minutes)+"분 "+str(total_seconds)+"초")
print("평균 인식 시간 : {:.2f}초".format(average_elapsed_time))
print("평균 정확도 : {:.2f}%".format(average_accuracy))

파일명 : 5350109-2001-0001-0001.jpg
인식 실행 중
정확도 : 40.27%
경과 시간 : 3.39초

파일명 : 5350109-2001-0001-0002.jpg
인식 실행 중
정확도 : 48.39%
경과 시간 : 1.24초

파일명 : 5350109-2001-0001-0004.jpg
인식 실행 중
정확도 : 32.24%
경과 시간 : 2.58초

파일명 : 5350109-2001-0001-0005.jpg
인식 실행 중
정확도 : 32.75%
경과 시간 : 2.92초

파일명 : 5350109-2001-0001-0006.jpg
인식 실행 중
정확도 : 47.37%
경과 시간 : 1.21초

파일명 : 5350109-2001-0001-0009.jpg
인식 실행 중
정확도 : 30.43%
경과 시간 : 2.71초

파일명 : 5350109-2001-0001-0010.jpg
인식 실행 중
정확도 : 33.59%
경과 시간 : 2.33초

파일명 : 5350109-2001-0001-0015.jpg
인식 실행 중
정확도 : 29.45%
경과 시간 : 2.55초

파일명 : 5350109-2001-0001-0016.jpg
인식 실행 중
정확도 : 31.82%
경과 시간 : 0.71초

파일명 : 5350109-2001-0001-0017.jpg
인식 실행 중
정확도 : 33.17%
경과 시간 : 3.03초

파일명 : 5350109-2001-0001-0018.jpg
인식 실행 중
정확도 : 27.22%
경과 시간 : 2.70초

파일명 : 5350109-2001-0001-0019.jpg
인식 실행 중
정확도 : 42.86%
경과 시간 : 0.77초

파일명 : 5350109-2001-0001-0020.jpg
인식 실행 중
정확도 : 29.26%
경과 시간 : 3.59초

파일명 : 5350109-2001-0001-0021.jpg
인식 실행 중
정확도 : 17.46%
경과 시간 : 2.03초

파일명 : 5350109-2001-0001-0022.jpg
인

정확도 : 37.50%
경과 시간 : 1.24초

파일명 : 5350109-2001-0001-0466.jpg
인식 실행 중
정확도 : 30.96%
경과 시간 : 2.96초

파일명 : 5350109-2001-0001-0467.jpg
인식 실행 중
정확도 : 28.90%
경과 시간 : 3.18초

파일명 : 5350109-2001-0001-0468.jpg
인식 실행 중
정확도 : 45.31%
경과 시간 : 1.19초

파일명 : 5350109-2001-0001-0469.jpg
인식 실행 중
정확도 : 34.81%
경과 시간 : 2.79초

파일명 : 5350109-2001-0001-0470.jpg
인식 실행 중
정확도 : 28.32%
경과 시간 : 2.57초

파일명 : 5350109-2001-0001-0471.jpg
인식 실행 중
정확도 : 38.10%
경과 시간 : 0.69초

파일명 : 5350109-2001-0001-0472.jpg
인식 실행 중
정확도 : 27.67%
경과 시간 : 3.02초

파일명 : 5350109-2001-0001-0473.jpg
인식 실행 중
정확도 : 36.23%
경과 시간 : 1.29초

파일명 : 5350109-2001-0001-0479.jpg
인식 실행 중
정확도 : 27.88%
경과 시간 : 3.15초

파일명 : 5350109-2001-0001-0480.jpg
인식 실행 중
정확도 : 48.28%
경과 시간 : 1.19초

파일명 : 5350109-2001-0001-0481.jpg
인식 실행 중
정확도 : 34.64%
경과 시간 : 2.74초

파일명 : 5350109-2001-0001-0482.jpg
인식 실행 중
정확도 : 30.54%
경과 시간 : 2.63초

파일명 : 5350109-2001-0001-0483.jpg
인식 실행 중
정확도 : 36.36%
경과 시간 : 0.71초

파일명 : 5350109-2001-0001-0484.jpg
인식 실행 중
정확도 : 18.18%
경과 시간 : 2.51초

파일명 : 

### 테스트 결과

1. 기본 값의 테스트 경우 약 평균 50~60% 대의 정확도가 나옴 (단순 값 비교 방식이어서 약간의 오차가 존재할 수 있음)

2. EasyOCR 속성을 다운그레이드 값으로 변경 시 기본 값 대비 평균 약 25% 차이 발생 (약 58% -> 32%)

3. GPU 환경에서 실행 시 CPU 환경 대비 평균 인식 소요 시간 약 7배 감소 (총 45분 -> 7분 / 평균 14초 -> 2초)